src/canister_tests/src/tests.rs

```diff
@@ -43,27 +40,184 @@ fn ii_canister_can_be_upgraded_and_rolled_back() {
+    #[test]
+    fn registration_with_mismatched_sender_fails() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        let challenge = api::create_challenge(&env, canister_id)?;
+        let result = api::register(
+            &env,
+            canister_id,
+            principal_2(),
+            &device_data_1(),
+            ChallengeAttempt {
+                chars: "a".to_string(),
+                key: challenge.challenge_key,
+            },
+        );
+
+        expect_user_error_with_message(
+            result,
+            CanisterCalledTrap,
+            Regex::new("[a-z\\d-]+ could not be authenticated against").unwrap(),
+        );
+        Ok(())
+    }
+
+    /// Tests that the solution to the captcha needs to be correct.
+    #[test]
+    fn should_not_allow_wrong_captcha() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+
+        let challenge = api::create_challenge(&env, canister_id)?;
+        let result = api::register(
+            &env,
+            canister_id,
+            principal_1(),
+            &device_data_1(),
+            ChallengeAttempt {
+                chars: "wrong solution".to_string(),
+                key: challenge.challenge_key,
+            },
+        )?;
+
+        assert!(matches!(result, RegisterResponse::BadChallenge));
+        Ok(())
+    }
+
+    /// Tests that there is a time limit for captchas.
+    /// Currently only checked by create_challenge, see L2-766.
+    #[test]
+    fn should_not_allow_expired_captcha() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+
+        let challenge = api::create_challenge(&env, canister_id)?;
+        env.advance_time(Duration::from_secs(301)); // one second longer than captcha validity
+
+        // required because register does not check captcha expiry
+        api::create_challenge(&env, canister_id)?;
+        let result = api::register(
+            &env,
+            canister_id,
+            principal_1(),
+            &device_data_1(),
+            ChallengeAttempt {
+                chars: "a".to_string(),
+                key: challenge.challenge_key,
+            },
+        )?;
+
+        assert!(matches!(result, RegisterResponse::BadChallenge));
+        Ok(())
+    }
+
+    /// Tests that there is a maximum number of captchas that can be created in a given timeframe.
+    #[test]
+    fn should_limit_captcha_creation() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+
+        for _ in 0..500 {
```

Comment:
This is probably too slow: https://github.com/dfinity/internet-identity/runs/7093616311?check_suite_focus=true#step:7:978

@nmattia : What do you think?

Reply:
Created ticket L2-791 to track this.